File: src/com/googlecode/utterlyidle/ApplicationStarter.scala
package org.webfabric.rest

import org.webfabric.servlet.{ContextPath, BasePath, WarRoot}
import javax.servlet.{ServletContext, ServletContextEvent, ServletContextListener}
import org.webfabric.properties.PropertiesApplication
import com.googlecode.yadic.SimpleContainer

class ApplicationStarter extends ServletContextListener{
  def createApplication(servletContext: ServletContext): Application = {
    val className = servletContext.getInitParameter(getClass.getName)
    if(className == null || className.equals("")){
      throw new UnsupportedOperationException("The web.xml must container a context-param called " + getClass.getName)
    }
    val aClass = Class.forName(className)
    val instance = new SimpleContainer().add(aClass).resolve(aClass)
    if(!instance.isInstanceOf[Application]){
      throw new UnsupportedOperationException(className + " must implement " + classOf[Application].getName )
    }
    instance.asInstanceOf[Application]
  }

  def contextInitialized(event: ServletContextEvent) = {
    val context = event.getServletContext
    val application = createApplication(context).addInstance(WarRoot(context)).addInstance(ContextPath(context))
    context.setAttribute(classOf[Application].getCanonicalName, application)
  }
  
  def contextDestroyed(event: ServletContextEvent) = {}
}

File: src/com/googlecode/utterlyidle/ApplicationStarter.java
package com.googlecode.utterlyidle;

import com.googlecode.utterlyidle.servlet.AttributeMap;
import com.googlecode.utterlyidle.servlet.ContextPath;
import com.googlecode.utterlyidle.servlet.WarRoot;
import com.googlecode.yadic.SimpleContainer;

import javax.servlet.ServletContext;
import javax.servlet.ServletContextEvent;
import javax.servlet.ServletContextListener;

import static com.googlecode.utterlyidle.servlet.AttributeMap.attributeMap;
import static com.googlecode.utterlyidle.servlet.ContextPath.contextPath;
import static com.googlecode.utterlyidle.servlet.WarRoot.warRoot;

public class ApplicationStarter implements ServletContextListener {
    private Application createApplication(ServletContext servletContext) {
        String className = servletContext.getInitParameter(getClass().getName());
        if (className == null || className.equals("")) {
            throw new UnsupportedOperationException("The web.xml must container a context-param called " + getClass().getName());
        }
        try {
            Class<?> aClass = Class.forName(className);
            Object instance = new SimpleContainer().add(aClass).resolve(aClass);
            if (!(instance instanceof Application)) {
                throw new UnsupportedOperationException(className + " must implement " + Application.class.getName());
            }
            return (Application) instance;
        } catch (ClassNotFoundException e) {
            throw new UnsupportedOperationException(e);
        }
    }

    public void contextInitialized(ServletContextEvent event) {
        ServletContext context = event.getServletContext();
        Application application = createApplication(context);
        application.applicationScope().
                addActivator(WarRoot.class, warRoot(context)).
                addActivator(ContextPath.class, contextPath(context)).
                addActivator(AttributeMap.class, attributeMap(context));
        context.setAttribute(Application.class.getCanonicalName(), application);
    }

    public void contextDestroyed(ServletContextEvent event) {
    }
}